ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,150 @@
+
+See TBDs
+
+## Summary
+
+This ship proposes means for Build users to define that their completed BuildRuns should be automatically cleaned up after some time or when a certain number of BuildRuns is reached.
+
+## Motivation
+
+In Shipwright, we separated the definition of a build and the run of a build into two artifacts, the Build and BuildRun. The BuildRun by its nature is a run-to-completion workload. Once it is completed, its only purpose is to view its status. Retaining the BuildRun for some time makes sense. In case of a successful BuildRun on finds there the digest of the image. For a failed BuildRun, one finds the error details and a pointer to the pod logs. The lifecycle of the BuildRun is coupled with the TaskRun and Pod using [owner references](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/).
+
+Over times, the number of BuildRuns in the system grows. This consumes resources in Kubernetes' database (usually etcd) and on the nodes (the pod logs).
+
+Eventually, users must manually clean up old BuildRuns which is not nice. Providing them means to define that BuildRuns get automatically cleaned up simplifies their lives.
+
+### Goals
+
+- Allow Build user to define how many completed BuildRuns should be kept.
+- Allow Build user to define how long a completed BuildRun should be kept.
+- Allow Build user to separate successful and failed BuildRuns in their cleanup handling.
+
+### Non-Goals
+
+- Allowing BuildRuns to be archived to an external database.
+
+## Proposal
+
+### API Changes
+
+We will extend the Build resource with an additional optional `retention` section`:
+
+```yaml
+spec:
+  retention:
+    ttlAfterFailed: duration
+    ttlAfterSuccessful: duration
+    successfulLimit: uint
+    failedLimit: uint
+```
+
+All fields inside the `retention` section are optional.
+
+The `ttlAfterFailed` and `ttlAfterSuccessful` fields holds a duration. They define the duration on how long a BuildRun should live after it failed, or successfully completed.
+
+A successful BuildRun is a BuildRun where the `type=Succeeded` condition is set to `"true"`. A failed BuildRun has the `type=Succeeded` condition is set to `"false"`, and contains therefore for example canceled BuildRuns as well.
+
+The two TTL fields behaves the same as the [`ttlSecondsAfterFinished` field for Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/). I decided against naming consistency with that feature, in favor of consistency inside our resource. Builds and BuildRuns already have a `timeout` field where we also use a duration while Kubernetes Jobs use an integer to store just seconds. Durations are easier to specify for users because they can simply write something like `6h` and don't need to calculate the amount of seconds.
+
+The `successfulLimit` and `failedLimit` fields hold a positive integer number. They define how many successful and failed BuildRuns for a Build can exist.
```

Comment:
My assumption is that if *either* the TTL or limit count is reached, pruning occurs.

But let's not assume it is obvious (which ever way you are thinking) and state that here, even if you make it clearer in the backend/details section.

By comparison, I would expect that distinction to be clear in the godoc of the API go files vs. having to pick up on that fact in the associated markdown files.